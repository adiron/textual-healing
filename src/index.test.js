const TextualHealing = require("./index.js");


const pizzaData = {
  "bread": [
    "sandwich",
    "pizza",
    "cake",
    "burrito",
  ],
  "kind": [
    "pepperoni",
    "veggie",
    "fish",
    "cheese",
    "chocolate",
    "disappointment",
  ],
  "start": [
    "{kind} {bread}",
  ],
};


test("instantiates", () => {
  const t = new TextualHealing({});
  expect(t).toBeInstanceOf(TextualHealing);
});

test("seed is properly set", () => {
  const s = 1234538;
  const t = new TextualHealing({}, s);
  expect(t.seed).toBe(s);
});

test("expands a simple token string", () => {
  const t = new TextualHealing({
    "start": [ "{a} {b}" ],
    "a": [ "a" ],
    "b": [ "b" ],
  });
  expect(t.start()).toBe("a b");
});

test("simple expansion (with random seed)", () => {
  const t = new TextualHealing(pizzaData, 1234830);

  const result = (new Array(100)).fill(null)
    .map(() => t.start())
    .join("\n");
  expect(result).toMatchSnapshot();
});

test("capitalize filter", () => {
  const t = new TextualHealing(pizzaData);
  expect(t.generate("{start:capitalize}").match(/^[A-Z]/)).toBeTruthy();
});

test("pluralize filter", () => {
  const t = new TextualHealing({
    "a": [ "sandwich" ],
    "b": [ "pizza" ],
    "c": [ "fly" ],
  });

  expect(t.generate("{a:pl}")).toBe("sandwiches");
  expect(t.generate("{b:pl}")).toBe("pizzas");
  expect(t.generate("{c:pl}")).toBe("flies");
});

test("definite filter", () => {
  const t = new TextualHealing({
    "a": [ "sandwich" ],
  });

  expect(t.generate("{a:definite}")).toBe("the sandwich");
});

test("indefinite filter", () => {
  const t = new TextualHealing({
    "a": [ "sandwich" ],
    "b": [ "hour" ],
  });

  expect(t.generate("{a:indefinite}")).toBe("a sandwich");
  expect(t.generate("{b:indefinite}")).toBe("an hour");
});

test("custom filters", () => {
  const t = new TextualHealing(
    {
      "a": [ "sandwich" ],
      "b": [ "hour" ],
    },
    undefined,
    {
      "uppercase": s => s.toUpperCase(),
    }
  );

  expect(t.generate("eating a {a:uppercase}")).toBe("eating a SANDWICH");
  expect(t.generate("{b:indefinite}")).toBe("hour"); // Graceful failure
});

test("multiple filters", () => {
  const t = new TextualHealing({
    "a": [ "sandwich" ],
  });

  expect(t.generate("{a:definite:pl:capitalize}")).toBe("The sandwiches");
  expect(t.generate("{a:capitalize:definite:pl}")).toBe("the Sandwiches");
});

test("union groups", () => {
  const t = new TextualHealing(pizzaData, 1234830);

  const result = (new Array(100)).fill(null)
    .map(() => t.generate("{kind|bread}"))
    .join("\n");
  expect(result).toMatchSnapshot();
});

test("error thrown for invalid groups", () => {
  const t = new TextualHealing(pizzaData);

  expect(() => {
    t.generate("{thisDoesntExist}");
  }).toThrow();
});

test("string returned as-is", () => {
  const t = new TextualHealing(pizzaData);
  expect(t.generate("hello")).toBe("hello");
});

test("tags do not alter patterns", () => {
  const t = new TextualHealing({
    "start": [ "{a (pizza)} {b (cake)}" ],
    "a": [ "a" ],
    "b": [ "b" ],
  });
  expect(t.start()).toBe("a b");
});

test("can tag and recall tags", () => {
  const t = new TextualHealing({
    "a": [ "disappointment" ],
  });
  const doubleString = t.generate("{a (0)} {(0)}");
  expect(doubleString).toBe("disappointment disappointment");
});

test("can tag and recall tags (advanced)", () => {
  const t = new TextualHealing(pizzaData, 123844);
  const pattern = "I like {kind (theKind)} {bread:pl (theBread)}. Although, I could trade {(theKind)} for {kind (newKind)}. That's right: {(newKind)} {(theBread)}.";
  const result = (new Array(100)).fill(null)
    .map(() => t.generate(pattern))
    .join("\n");
  expect(result).toMatchSnapshot();
});
