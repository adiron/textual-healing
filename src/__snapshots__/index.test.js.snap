// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple expansion (with random seed) 1`] = `
"fish burrito
fish cake
chocolate burrito
pepperoni pizza
fish sandwich
fish burrito
veggie cake
fish burrito
pepperoni burrito
fish sandwich
pepperoni pizza
cheese burrito
chocolate sandwich
cheese cake
fish pizza
cheese sandwich
cheese pizza
disappointment cake
fish pizza
veggie cake
cheese pizza
fish cake
chocolate sandwich
disappointment sandwich
fish burrito
pepperoni sandwich
disappointment burrito
chocolate sandwich
fish sandwich
pepperoni sandwich
veggie sandwich
pepperoni cake
chocolate cake
veggie sandwich
disappointment burrito
pepperoni burrito
fish burrito
pepperoni burrito
veggie pizza
cheese burrito
chocolate pizza
pepperoni cake
fish sandwich
cheese pizza
fish cake
fish sandwich
disappointment cake
chocolate sandwich
fish pizza
disappointment cake
chocolate burrito
pepperoni burrito
veggie burrito
pepperoni pizza
fish cake
fish pizza
chocolate pizza
veggie cake
chocolate pizza
disappointment cake
fish burrito
disappointment sandwich
cheese pizza
cheese pizza
chocolate pizza
veggie sandwich
veggie burrito
pepperoni sandwich
disappointment cake
chocolate burrito
disappointment sandwich
chocolate pizza
chocolate cake
fish pizza
veggie sandwich
veggie cake
veggie sandwich
disappointment cake
disappointment cake
disappointment sandwich
disappointment sandwich
veggie pizza
disappointment sandwich
veggie sandwich
cheese pizza
chocolate pizza
fish sandwich
cheese burrito
disappointment burrito
pepperoni cake
chocolate cake
pepperoni pizza
veggie cake
pepperoni burrito
fish cake
veggie burrito
fish cake
pepperoni burrito
disappointment burrito
disappointment cake"
`;

exports[`union groups 1`] = `
"cheese
chocolate
cake
veggie
cheese
pizza
fish
sandwich
burrito
disappointment
veggie
cheese
veggie
cheese
fish
cake
cheese
cake
chocolate
fish
disappointment
chocolate
cheese
burrito
cheese
pepperoni
pizza
pizza
chocolate
pepperoni
burrito
pepperoni
cheese
chocolate
disappointment
burrito
cheese
pizza
pepperoni
cake
disappointment
sandwich
pizza
chocolate
cheese
chocolate
disappointment
pepperoni
pizza
sandwich
chocolate
cake
burrito
disappointment
veggie
chocolate
cheese
pepperoni
cheese
pizza
veggie
disappointment
chocolate
fish
cheese
sandwich
sandwich
pizza
veggie
sandwich
burrito
veggie
disappointment
cheese
cake
chocolate
pepperoni
pepperoni
chocolate
cake
pizza
sandwich
pizza
pepperoni
fish
chocolate
fish
sandwich
pepperoni
fish
pizza
fish
pepperoni
sandwich
pepperoni
disappointment
pizza
pizza
pizza
sandwich"
`;
